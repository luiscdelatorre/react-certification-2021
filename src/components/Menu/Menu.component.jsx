import React from 'react';
import {
  IoEnterOutline,
  IoExitOutline,
  IoHomeOutline,
  IoOptions,
  IoStar,
} from 'react-icons/io5';
import { useHistory } from 'react-router';
import { useAuth } from '../../providers/Auth';
import { useOptions } from '../../providers/Options/Options.provider';
import { MenuContainer, MenuIcon, MenuItem, MenuLink, MenuList } from './Menu.styles';

const Menu = () => {
  const history = useHistory();
  const { authenticated, logout } = useAuth();
  const { state } = useOptions();

  const deAuthenticate = (event) => {
    event.preventDefault();
    logout();
    history.push('/');
  };

  return (
    <MenuContainer data-testid="menu">
      <MenuList>
        <MenuItem>
          <MenuLink to="/" exact>
            <MenuIcon isExpanded={!state.menuCompact}>
              <IoHomeOutline />
              <span>Home</span>
            </MenuIcon>
          </MenuLink>
        </MenuItem>
        <MenuItem disabled>
          <MenuIcon isExpanded={!state.menuCompact}>
            <IoStar />
            <span>Favorites</span>
          </MenuIcon>
        </MenuItem>
        <MenuItem disabled>
          <MenuIcon isExpanded={!state.menuCompact}>
            <IoOptions />
            <span>Options</span>
          </MenuIcon>
        </MenuItem>
        <MenuItem>
          {authenticated ? (
            <MenuLink
              to="/login"
              exact
              onClick={deAuthenticate}
              data-testid="button-logout"
            >
              <MenuIcon isExpanded={!state.menuCompact}>
                <IoExitOutline />
                <span>Logout</span>
              </MenuIcon>
            </MenuLink>
          ) : (
            <MenuLink to="/login" exact>
              <MenuIcon isExpanded={!state.menuCompact}>
                <IoEnterOutline />
                <span>Login</span>
              </MenuIcon>
            </MenuLink>
          )}
        </MenuItem>
      </MenuList>
    </MenuContainer>
  );
};

export default Menu;
