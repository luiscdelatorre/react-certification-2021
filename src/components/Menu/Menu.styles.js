import { NavLink } from 'react-router-dom';
import styled from 'styled-components';
import colors from '../../config/colors';
import device from '../../config/device';

const MenuContainer = styled.nav`
  grid-area: menu;
  background-color: ${(props) => props.theme.menuBackground};
  padding: 1rem;
  border-top: 1px solid ${(props) => props.theme.borderColor};
  @media ${device.laptop} {
    border-top: 0;
    border-right: 1px solid ${(props) => props.theme.borderColor};
    padding: 2rem;
  }
`;

const MenuList = styled.ul`
  display: flex;
  justify-content: space-around;
  height: 100%;
  @media ${device.laptop} {
    flex-direction: column;
    justify-content: flex-start;
    border-bottom: 1px solid ${(props) => props.theme.borderColor};
    height: auto;
  }
`;

const MenuItem = styled.li`
  color: ${(props) => props.theme.textSecondary};
  padding: 0.5rem;
  width: 5rem;
  height: 5rem;
  margin-bottom: 1rem;
  position: relative;
  z-index: 0;
  @media ${device.laptop} {
    width: 100%;
  }
`;

const MenuLink = styled(NavLink)`
  color: inherit;
  text-decoration: none;
  width: 100%;
  height: 100%;

  &[disabled] {
    cursor: not-allowed;
  }

  &::before {
    content: '';
    background-color: transparent;
    border-radius: ${(props) => props.theme.borderRadiusLg};
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: -1;
    transition: ${(props) => props.theme.transitionDefault};
  }

  &:hover:not(.active) {
    color: ${colors.gray100};
    &::before {
      background-color: ${(props) => props.theme.primary};
      animation-name: menu-button-animation;
      animation-timing-function: ease-in;
      animation-duration: 300ms;
    }
  }
  &.active {
    color: ${colors.gray100};
    &::before {
      background-color: ${(props) => props.theme.primary};
    }
  }
`;

const MenuIcon = styled.div`
  display: flex;
  align-items: center;
  font-size: 2rem;
  height: 100%;
  flex-direction: column;
  justify-content: space-around;
  @media ${device.laptop} {
    flex-direction: ${(props) => (props.isExpanded ? 'row' : 'column')};
    justify-content: ${(props) => (props.isExpanded ? 'flex-start' : 'space-around')};
    padding: ${(props) => (props.isExpanded ? '1.5rem' : '0')};
  }
  ${MenuItem}:hover & {
    color: ${(props) => props.theme.white};
  }
  span {
    display: inline-block;
    font-size: 1rem;
    @media ${device.laptop} {
      font-size: ${(props) => (props.isExpanded ? '1.4rem' : '1rem')};
      margin-left: ${(props) => (props.isExpanded ? '2rem' : '0')};
    }
  }
`;

export { MenuContainer, MenuList, MenuItem, MenuLink, MenuIcon };
