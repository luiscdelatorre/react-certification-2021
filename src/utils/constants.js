const AUTH_STORAGE_KEY = 'wa_cert_authenticated';
const SEARCH_STORAGE_KEY = 'wa_cert_search';
const API_URL = `https://www.googleapis.com/youtube/v3`;

export { AUTH_STORAGE_KEY, SEARCH_STORAGE_KEY, API_URL };
