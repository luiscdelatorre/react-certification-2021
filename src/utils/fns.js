const random = (limit) => {
  return Math.floor(Math.random() * limit);
};

const formatDate = (dateString) => {
  const date = new Date(dateString);
  return date.toLocaleDateString();
};

export { random, formatDate };
