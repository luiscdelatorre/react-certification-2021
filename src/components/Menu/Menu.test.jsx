import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import AuthProvider from '../../providers/Auth';
import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';
import Menu from './Menu.component';

describe('Menu Component', () => {
  it('Should Render all Menu options', () => {
    render(
      <BrowserRouter>
        <AuthProvider>
          <Menu isExpanded={false} />
        </AuthProvider>
      </BrowserRouter>
    );
    const totalMenuItems = screen.getAllByRole('listitem').length;

    expect(totalMenuItems).toBe(4);
    expect(screen.getByText('Home')).toBeInTheDocument();
    expect(screen.getByText('Favorites')).toBeInTheDocument();
    expect(screen.getByText('Options')).toBeInTheDocument();
    expect(screen.getByText('Login')).toBeInTheDocument();
  });

  it('Should trigger logout event', () => {
    storage.set(AUTH_STORAGE_KEY, true);

    render(
      <BrowserRouter>
        <AuthProvider>
          <Menu isExpanded={false} />
        </AuthProvider>
      </BrowserRouter>
    );
    expect(screen.queryByText('Login')).not.toBeInTheDocument();
    expect(screen.queryByText('Logout')).toBeInTheDocument();

    const button = screen.getByTestId('button-logout');
    fireEvent.click(button);

    expect(screen.queryByText('Login')).toBeInTheDocument();
    expect(screen.queryByText('Logout')).not.toBeInTheDocument();
  });

  it('Validates Menu snapshot', () => {
    const component = renderer.create(
      <BrowserRouter>
        <AuthProvider>
          <Menu isExpanded={false} />
        </AuthProvider>
      </BrowserRouter>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
