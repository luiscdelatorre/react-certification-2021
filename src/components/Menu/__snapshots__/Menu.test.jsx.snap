// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Component Validates Menu snapshot 1`] = `
<nav
  className="sc-bdfBwQ hEEpsM"
  data-testid="menu"
>
  <ul
    className="sc-gsTCUz dLjuc"
  >
    <li
      className="sc-dlfnbm cvUCAY"
    >
      <a
        aria-current="page"
        className="sc-hKgILt kmoA-DQ active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        <div
          className="sc-eCssSg hMCHXn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M80 212v236a16 16 0 0016 16h96V328a24 24 0 0124-24h80a24 24 0 0124 24v136h96a16 16 0 0016-16V212"
              fill="none"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="32"
            />
            <path
              d="M480 256L266.89 52c-5-5.28-16.69-5.34-21.78 0L32 256m368-77V64h-48v69"
              fill="none"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="32"
            />
          </svg>
          <span>
            Home
          </span>
        </div>
      </a>
    </li>
    <li
      className="sc-dlfnbm cvUCAY"
      disabled={true}
    >
      <div
        className="sc-eCssSg hMCHXn"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z"
          />
        </svg>
        <span>
          Favorites
        </span>
      </div>
    </li>
    <li
      className="sc-dlfnbm cvUCAY"
      disabled={true}
    >
      <div
        className="sc-eCssSg hMCHXn"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M64 144h226.75a48 48 0 0090.5 0H448a16 16 0 000-32h-66.75a48 48 0 00-90.5 0H64a16 16 0 000 32zm384 224h-66.75a48 48 0 00-90.5 0H64a16 16 0 000 32h226.75a48 48 0 0090.5 0H448a16 16 0 000-32zm0-128H221.25a48 48 0 00-90.5 0H64a16 16 0 000 32h66.75a48 48 0 0090.5 0H448a16 16 0 000-32z"
          />
        </svg>
        <span>
          Options
        </span>
      </div>
    </li>
    <li
      className="sc-dlfnbm cvUCAY"
    >
      <a
        aria-current={null}
        className="sc-hKgILt kmoA-DQ"
        href="/login"
        onClick={[Function]}
      >
        <div
          className="sc-eCssSg hMCHXn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M176 176v-40a40 40 0 0140-40h208a40 40 0 0140 40v240a40 40 0 01-40 40H216a40 40 0 01-40-40v-40"
              fill="none"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="32"
            />
            <path
              d="M272 336l80-80-80-80M48 256h288"
              fill="none"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="32"
            />
          </svg>
          <span>
            Login
          </span>
        </div>
      </a>
    </li>
  </ul>
</nav>
`;
